fv3core/fv3core/initialization/baroclinic.py

```diff
@@ -0,0 +1,506 @@
+            grid_vector_component[:, :, 1] * np.cos(lon)
+            - grid_vector_component[:, :, 0] * np.sin(lon)
+        )[:, :, None]
+    )
+
+
+def wind_component_calc(
+    shape,
+    eta_v,
+    lon,
+    lat,
+    grid_vector_component,
+    islice,
+    islice_grid,
+    jslice,
+    jslice_grid,
+):
+    slice_grid = (islice_grid, jslice_grid)
+    slice_3d = (islice, jslice, slice(None))
+    u_component = np.zeros(shape)
+    u_component[slice_3d] = jablo_init.baroclinic_perturbed_zonal_wind(
+        eta_v, lon[slice_grid], lat[slice_grid]
+    )
+    u_component[slice_3d] = local_coordinate_transformation(
+        u_component[slice_3d],
+        lon[slice_grid],
+        grid_vector_component[islice_grid, jslice_grid, :],
+    )
+    return u_component
+
+
+def initialize_zonal_wind(
+    u,
+    eta,
+    eta_v,
+    lon,
+    lat,
+    east_grid_vector_component,
+    center_grid_vector_component,
+    islice,
+    islice_grid,
+    jslice,
+    jslice_grid,
+    axis,
+):
+    shape = u.shape
+    uu1 = wind_component_calc(
+        shape,
+        eta_v,
+        lon,
+        lat,
+        east_grid_vector_component,
+        islice,
+        islice,
+        jslice,
+        jslice_grid,
+    )
+    uu3 = wind_component_calc(
+        shape,
+        eta_v,
+        lon,
+        lat,
+        east_grid_vector_component,
+        islice,
+        islice_grid,
+        jslice,
+        jslice,
+    )
+    upper = (slice(None),) * axis + (slice(0, -1),)
+    lower = (slice(None),) * axis + (slice(1, None),)
+    pa1, pa2 = lon_lat_midpoint(lon[upper], lon[lower], lat[upper], lat[lower], np)
+    uu2 = wind_component_calc(
+        shape,
+        eta_v,
+        pa1,
+        pa2,
+        center_grid_vector_component,
+        islice,
+        islice,
+        jslice,
+        jslice,
+    )
+    u[islice, jslice, :] = 0.25 * (uu1 + 2.0 * uu2 + uu3)[islice, jslice, :]
+
+
+def compute_grid_edge_midpoint_latitude_components(lon, lat):
+    _, lat_avg_x_south = lon_lat_midpoint(
+        lon[0:-1, :], lon[1:, :], lat[0:-1, :], lat[1:, :], np
+    )
+    _, lat_avg_y_east = lon_lat_midpoint(
+        lon[1:, 0:-1], lon[1:, 1:], lat[1:, 0:-1], lat[1:, 1:], np
+    )
+    _, lat_avg_x_north = lon_lat_midpoint(
+        lon[0:-1, 1:], lon[1:, 1:], lat[0:-1, 1:], lat[1:, 1:], np
+    )
+    _, lat_avg_y_west = lon_lat_midpoint(
+        lon[:, 0:-1], lon[:, 1:], lat[:, 0:-1], lat[:, 1:], np
+    )
+    return lat_avg_x_south, lat_avg_y_east, lat_avg_x_north, lat_avg_y_west
+
+
+def cell_average_nine_point(pt1, pt2, pt3, pt4, pt5, pt6, pt7, pt8, pt9):
+    """
+    9-point average: should be 2nd order accurate for a rectangular cell
+    9  4  8
+    5  1  3
+    6  2  7
+    """
+    return (
+        0.25 * pt1 + 0.125 * (pt2 + pt3 + pt4 + pt5) + 0.0625 * (pt6 + pt7 + pt8 + pt9)
+    )
+
+
+def cell_average_nine_components(
+    component_function,
+    component_args,
+    lon,
+    lat,
+    lat_agrid,
+    grid_slice,
```

Comment:
The type of this argument is non-intuitive and would be particularly helpful to type hint.
```suggestion
    grid_slice: Tuple[slice, slice],
```

Reply:
If you do still want to pull `grid_slice` up one level, the way you would do it is by passing in `lon` and `lat` sliced on their own compute domains - that is, up to nx+1 and ny+1 since they're defined on cell corners. Then below, every access to lat or lon should be clipping off 1 point. pt6 becomes `lat[:-1, :-1]` while pt9 becomes `lat[:-1, 1:]`.